src/include/sof/audio/volume.h

```diff
@@ -231,6 +234,71 @@ static inline void peak_vol_update(struct vol_data *cd)
 #endif
 }
 
+/**
+ * \brief Ramps volume changes over time.
+ * \param[in,out] mod Volume processing module handle
+
+ */
+static inline void volume_ramp(struct vol_data *cd)
+{
+	int32_t new_vol;
+	int32_t ramp_time;
+	int32_t tvolume;
+	int32_t volume;
+	int i;
+	bool ramp_finished = true;
+
+	cd->copy_gain = true;
+
+	/* The first is set and cleared to indicate ongoing ramp, the
+	 * latter is set once to enable self launched ramp only once
+	 * in stream start.
+	 */
+	cd->vol_ramp_active = true;
+
+	/* Current ramp time in Q29.3 milliseconds. Note that max. ramp length
+	 * can be 1.3s at 192 kHz rate and 5.5s at 48 kHz rate without
+	 * exceeding int32_t range. Inverse of sample rate is 1000/sample_rate
+	 * for milliseconds.
+	 */
+	ramp_time = Q_MULTSR_32X32((int64_t)cd->vol_ramp_elapsed_frames,
+				   cd->sample_rate_inv, 0, 31, 3);
+
+	/* Update each volume if it's not at target for active channels */
+	for (i = 0; i < cd->channels; i++) {
+		/* skip if target reached */
+		volume = cd->volume[i];
+		tvolume = cd->tvolume[i];
+		if (volume == tvolume)
+			continue;
+
+		/* Update volume gain with ramp. The ramp gain value is
+		 * calculated from previous gain and ramp time. The slope
+		 * coefficient is calculated in volume_set_chan().
+		 */
+		new_vol = cd->ramp_func(cd, ramp_time, i);
+		if (volume < tvolume) {
+			/* ramp up, check if ramp completed */
+			if (new_vol >= tvolume)
+				new_vol = tvolume;
+			else
+				ramp_finished = false;
+		} else {
+			/* ramp down */
+			if (new_vol <= tvolume)
+				new_vol = tvolume;
+			else
+				ramp_finished = false;
+		}
+		cd->volume[i] = new_vol;
+	}
+
+	if (ramp_finished) {
```

Comment:
ramp_finished is for all channel, however, setting is based on per channel, i.e, as long as one channel  need ramp, ramp is happening, is it expected with design?

Reply:
The ramp channels loop function is called until every channel has completed. However typically the channel gains are the same value so there is not much help if try to optimize the single channel update. Also finding similar channels for 2ch might not help much. With larger channels count maybe.